### ISSUE TITLE
Feature: output in SARIF format 

### ISSUE THREAD

**Issue by htadashi** [trailofbits/pip-audit#206]
**Is your feature request related to a problem? Please describe.**

No.

**Describe the solution you'd like**

It would be nice to have the option of getting pip-audit's output in [SARIF](https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/sarif-support-for-code-scanning) format. This would help create GitHub actions (#198) that naturally integrate with GitHub security advisories page.

**Comment by woodruffw** [comment]
Thanks for the request!

Looks like this is the reference SARIF implementation for Python: https://github.com/microsoft/sarif-python-om

**Comment by artemrys** [comment]
@woodruffw I would like to take this one, especially since we have an official GitHub Action already implemented.

**Comment by woodruffw** [comment]
@artemrys Please do! Feel free to ping me if you have any technical questions about the codebase.

**Comment by artemrys** [comment]
Started working on this feature, created a sample SARIF file, and uploaded it to GitHub, attaching the screenshot of how it looks.

Some links for future reference:

* tool created by Microsoft to validate SARIF files [here](https://sarifweb.azurewebsites.net/Validation)
* found an example of a tool that performs similar actions to what `pip-audit` do and we can see how they are assembling the SARIF file [here](https://github.com/redhat-actions/crda/blob/b0a4d3b7adb8d7778698ef79d5b0ccbbc36607f6/src/convert.ts#L306)

I'll continue working on exploring more options.

<img width="956" alt="image" src="https://user-images.githubusercontent.com/11015487/201205356-9be00022-2cab-41aa-b557-10e247a69d32.png">

**Comment by artemrys** [comment]
I think we might be blocked here because GitHub requires a [`physicalLocation`](https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/sarif-support-for-code-scanning#physicallocation-object) field to be populated in a SARIF file.

The `physicalLocation` requires the file location, line, and column where the code scanning result was found. If I understand correctly, `pip-audit` currently does not provide this information in the `VulnerabilityResult` class.

Especially, it hurts with a vulnerability in a transitive dependency. If the requirements file contains only`pytest==7.1.3` line and `pytest v7.1.3` has a dependency on `py` which has a vulnerability (PYSEC-2022-42969), then showing the exact place in the "code" with a vulnerability might be tricky.

**Comment by woodruffw** [comment]
> The `physicalLocation` requires the file location, line, and column where the code scanning result was found. If I understand correctly, `pip-audit` currently does not provide this information in the `VulnerabilityResult` class.

Yeah, we don't currently provide this in `VulnerabilityResult`, so we'd need to. Importantly, we *can't* provide it for every auditing "source": environments, for example, don't come from a file input so they don't have `(file, line, col)` information.

I'm not sure if this is the *right* approach, but for transitive dependencies we could probably flag them based on the file location of the direct dependencies that include them. We might need to think a bit more about the UX there to clarify that that's what we're doing, but I think that's the best we can do in terms of surfacing them.

**Comment by woodruffw** [comment]
To summarize: I'd be okay with us adding a `source_locations: Optional[List[SourceLocation]]` to each `Dependency`, which `VulnerabiliyResult` would then map with.

`SourceLocation` would then be:

```py
class SourceLocation:
    path: Path
    line: int
    col: int
```

...and `source_locations` needs to be a list, because of subdependencies: subdependency `S` might be introduced by deps `A` and `B`, so we want to record the source locations of *both* `A` and `B`.

**Comment by jsoref** [comment]
If you generate a sarif file and it's slightly incomplete, but some items have a physical location and you also generate an artifact, then people can have most of your cake and eat it too.

That's been my approach to sarif integration w/ github. Don't let the perfect be the enemy of the good.

**Comment by woodruffw** [comment]
I don't think anybody is letting the perfect be the enemy of the good here; if you'd like to see this feature added in a state that doesn't include source locations, please go ahead and work on it!

(I'm happy to merge SARIF support in whatever state we can, so long as (1) we conform to the schema's requirements, and (2) test it as comprehensively as the other output formats.)

**Comment by denishpillay** [comment]
Any updates on this feature request? More and more security visualization tools use SARIF now. Would be great to have it for pip-audit

**Comment by woodruffw** [comment]
Nope; this is still waiting for someone who is interested in taking it on.

Please don't take this the wrong way, but you almost never need to ask whether there's been an update or not: if the issue is open and hasn't received an update since the last activity, then there is no update.

**Comment by woodruffw** [comment]
@omobolajiadeyan Please do not submit LLM-generated comments or PRs to this project. They will not be reviewed.

**Comment by cheesymoon** [comment]
Not native support, but until that lands: https://github.com/sarif-kit/sarif-kit converts `pip-audit -f json` to SARIF 2.1.0. CLI on PyPI, or a container action:

```yaml
- run: pipx run pip-audit -r requirements.txt -f json > pip-audit.json || true
- uses: sarif-kit/sarif-kit@v0.1.1
  with:
    tool: pip-audit
    input: pip-audit.json
    output: pip-audit.sarif
    dep-file: requirements.txt
- uses: github/codeql-action/upload-sarif@v4
  with:
    sarif_file: pip-audit.sarif
    category: pip-audit
```

The `|| true` is because pip-audit exits nonzero on findings. It won't mask a dead audit though, pip-audit writes JSON even on a clean run, so an empty file fails the convert step.

Re the physicalLocation discussion from 2022: findings point at the manifest (`dep-file`), transitive ones too, since that's the file you'd edit anyway. GitHub takes the whole-file location fine. Each advisory gets its own rule, title comes from the advisory summary, CVE/GHSA aliases and fix versions go in the message.

Docs: https://github.com/sarif-kit/sarif-kit/blob/master/docs/pip-audit.md

**Comment by omobolajiadeyan** [comment]
Thanks for sharing this workaround and the detailed Actions example. `sarif-kit` looks useful for bridging the gap in the meantime, especially with the manifest location and per-advisory rule mapping. Native SARIF output in `pip-audit` would still be valuable, but this gives users a practical path today.
